File: 0036-valid-sudoku/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int rows[9] = {0};
        int cols[9] = {0};
        int boxes[9] = {0};
        
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                char val = board[r][c];
                if (val == '.') {
                    continue;
                }
                
                int digit = val - '1';
                int mask = 1 << digit;
                int boxIdx = (r / 3) * 3 + (c / 3);
                
                // Check duplication
                if ((rows[r] & mask) != 0 || (cols[c] & mask) != 0 || (boxes[boxIdx] & mask) != 0) {
                    return false;
                }
                
                // Mark as seen
                rows[r] |= mask;
                cols[c] |= mask;
                boxes[boxIdx] |= mask;
            }
        }
        
        return true;
    }
};

File: 0036-valid-sudoku/solution.py
class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9
        
        for r in range(9):
            for c in range(9):
                val = board[r][c]
                if val == '.':
                    continue
                    
                digit = int(val) - 1
                mask = 1 << digit
                box_idx = (r // 3) * 3 + (c // 3)
                
                # Check duplication
                if (rows[r] & mask) or (cols[c] & mask) or (boxes[box_idx] & mask):
                    return False
                    
                # Mark as seen
                rows[r] |= mask
                cols[c] |= mask
                boxes[box_idx] |= mask
                
        return True
